components/omega/src/base/MachEnv.cpp

```diff
@@ -0,0 +1,441 @@
+      NewEnv.NumTasks   = -999;
+   }
+
+   // Set task 0 as master
+   NewEnv.MasterTask = 0;
+
+   // determine if this task is the master task
+   if (NewEnv.MyTask == NewEnv.MasterTask) {
+      NewEnv.MasterTaskFlag = true;
+   } else {
+      NewEnv.MasterTaskFlag = false;
+   }
+
+#ifdef OMEGA_THREADED
+   // total number of OpenMP threads
+   NewEnv.NumThreads = omp_get_num_threads();
+#else
+   NewEnv.NumThreads = 1;
+#endif
+
+   // Add to list of environments
+   OMEGA::MachEnv::addEnv(Name, &NewEnv);
+
+} // end constructor with selected tasks
+
+// Remove/delete functions
+//------------------------------------------------------------------------------
+// Remove environment
+
+void MachEnv::removeEnv(const std::string name // [in] name of env to remove
+) {
+
+   AllEnvs.erase(name);
+
+} // end removeEnv
+
+// Retrieval functions
+//------------------------------------------------------------------------------
+// Get default environment
+MachEnv *MachEnv::getDefaultEnv() { return MachEnv::DefaultEnv; }
+
+//------------------------------------------------------------------------------
+// Get environment by name
+MachEnv *MachEnv::getEnv(const std::string name ///< [in] name of environment
+) {
+
+   // look for an instance of this name
+   auto it = AllEnvs.find(name);
+
+   // if found, return the environment pointer
+   if (it != AllEnvs.end()) {
+      return it->second;
+
+      // otherwise print an error and return a null pointer
+   } else {
+      std::cerr << "Attempt to retrieve a non-existent MachEnv " << name
+                << " has not been defined or has been removed";
+      return nullptr;
+   }
+
+} // end getEnv
+
+//------------------------------------------------------------------------------
+// Get communicator for an environment
+MPI_Comm MachEnv::getComm() const { return Comm; }
+
+//------------------------------------------------------------------------------
+// Get local task/rank ID
+int MachEnv::getMyTask() const { return MyTask; }
+
+//------------------------------------------------------------------------------
+// Get total number of MPI tasks/ranks
+int MachEnv::getNumTasks() const { return NumTasks; }
+
+//------------------------------------------------------------------------------
+// Get task ID for the master task (typically 0)
+int MachEnv::getMasterTask() const { return MasterTask; }
+
+//------------------------------------------------------------------------------
+// Determine whether local task is the master
+
+bool MachEnv::isMasterTask() const { return MasterTaskFlag; }
+
+//------------------------------------------------------------------------------
+// Determine whether local task is in this communicator's group
+
+bool MachEnv::isMember() const { return MemberFlag; }
+
+//------------------------------------------------------------------------------
+// Set task ID for the master task (if not 0)
+
+int MachEnv::setMasterTask(const int TaskID) {
```

Comment:
I guess the expectation is that we can dynamically change master even after comm creation. Would be great for adaptive/dynamic load balancing if we can make it work throughout the model.

Reply:
Yup, this could be part of load-balancing, especially if the default master is particularly over-loaded.